### PR TITLE
fix: type coercion support date - date

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #.

# Rationale for this change

# What changes are included in this PR?

type coercion support date - date

# Are these changes tested?

need arrow-rs support it, and then we add slt to test it.

# Are there any user-facing changes?

### CODE REVIEW

**Comment by jackwener** [comment]
wait for arrow-rs support `date` - `date`

File: datafusion/core/tests/sqllogictests/test_files/dates.slt [hunk 1]
```diff
@@ -90,7 +90,7 @@ select i_item_desc from test
 where d3_date > now() + '5 days';
 
 # DATE minus DATE
-query error DataFusion error: Error during planning: Unsupported argument types\. Can not evaluate Date32 \- Date32
+query error DataFusion error: Optimizer rule 'simplify_expressions' failed\ncaused by\nArrow error: Cast error: Cannot perform arithmetic operation between array of type Date32 and array of type Date32
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/dates.slt:93]: I filed https://github.com/apache/arrow-rs/issues/4383 to track arrow kernel support

```suggestion
# https://github.com/apache/arrow-rs/issues/4383
query error DataFusion error: Optimizer rule 'simplify_expressions' failed\ncaused by\nArrow error: Cast error: Cannot perform arithmetic operation between array of type Date32 and array of type Date32
```

File: datafusion/expr/src/type_coercion/binary.rs [hunk 2]
```diff
@@ -738,6 +740,8 @@ fn temporal_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataTyp
         (Interval(YearMonth), Interval(YearMonth)) => Some(Interval(YearMonth)),
         (Interval(DayTime), Interval(DayTime)) => Some(Interval(DayTime)),
         (Interval(_), Interval(_)) => Some(Interval(MonthDayNano)),
+        (Date32, Date32) => Some(Date32),
```
Comment (alamb) [inline datafusion/expr/src/type_coercion/binary.rs:743]: Maybe we should put a generic clause that if `lhs_type == rhs_type` we can return that type?

**Review by alamb — APPROVED** [review]
LGTM -- thanks @jackwener

File: datafusion/expr/src/type_coercion/binary.rs (hunk 2 above)
Comment (jackwener) [reply datafusion/expr/src/type_coercion/binary.rs]: Yes!

**Comment by jackwener** [comment]
Thanks @alamb review

File: datafusion/expr/src/type_coercion/binary.rs [hunk 3]
```diff
@@ -733,10 +735,13 @@ fn is_time_with_valid_unit(datatype: DataType) -> bool {
 fn temporal_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataType> {
     use arrow::datatypes::DataType::*;
     use arrow::datatypes::IntervalUnit::*;
+    use arrow::datatypes::TimeUnit::*;
+
+    if lhs_type == rhs_type {
```
Comment (alamb) [inline datafusion/expr/src/type_coercion/binary.rs:740]: 👍

**Pull request merged**